File: algorithms/code/hackerrank/unbounded_knapsack/unbounded_knapsack.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/unbounded-knapsack

import io
import os
import sys
import unittest


def unboundedKnapsack(w, values):
    weights = values  # Special case
    n = len(values)
    m = [0] * (w + 1)
    for wi in range(w + 1):
        for vi in range(n):
            if weights[vi] <= wi:
                m[wi] = max(m[wi], m[wi - weights[vi]] + values[vi])
    return m[w]


def main():
    fptr = sys.stdout
    t = int(input().strip())
    for _ in range(t):
        first_multiple_input = input().rstrip().split()
        _ = int(first_multiple_input[0])  # n
        k = int(first_multiple_input[1])
        arr = list(map(int, input().rstrip().split()))
        result = unboundedKnapsack(k, arr)
        fptr.write(str(result) + '\n')


if __name__ == '__main__':  # pragma: no cover
    main()


class TestCode(unittest.TestCase):
    def generalized_test(self, which):
        resources = os.path.dirname(__file__)
        with (
            open(
                os.path.join(resources, f'output{which}.txt'),
                'r',
                encoding='utf-8',
            ) as expected,
            open(
                os.path.join(resources, f'input{which}.txt'),
                'r',
                encoding='utf-8',
            ) as sys.stdin,
            io.StringIO() as sys.stdout,
        ):
            main()
            self.assertEqual(sys.stdout.getvalue(), expected.read())

    def test_example(self):
        self.generalized_test('example')

File: algorithms/code/hackerrank/unbounded_knapsack/unbounded_knapsack.rs
// https://www.hackerrank.com/challenges/unbounded-knapsack

use crate::hackerrank::tester::{read_input, write_and_check_output};
use std::cmp::max;

pub fn unbounded_knapsack(k: i32, arr: &[i32]) -> i32 {
    let n = arr.len();
    let mut m = vec![0; k as usize + 1];
    for wi in 0..=k {
        for vi in 0..n {
            if arr[vi] <= wi {
                m[wi as usize] = max(m[wi as usize], m[(wi - arr[vi]) as usize] + arr[vi]);
            }
        }
    }
    m[k as usize]
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let input = read_input(origin, name);
    let t = input[0][0].parse::<i32>().unwrap();
    let mut results: Vec<String> = Vec::new();
    let mut offset = 1;
    for _ in 0..t {
        let k = input[offset][1].parse::<i32>().unwrap();
        let arr: Vec<i32> = input[offset + 1]
            .iter()
            .map(|s| s.to_string().parse::<i32>().unwrap())
            .collect();
        offset += 2;
        let result = unbounded_knapsack(k, &arr);
        results.push(result.to_string());
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("example");
    }
}
